### DISCUSSION TITLE
Exception contracts

### DISCUSSION THREAD

**Discussion by vsfeedback** [dotnet/csharplang#3844] [General]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/content/idea/940560/exception-contracts.html)._

---
<p>In a similar way to code contracts allow an attribute based way to handle exceptions in a method. Often time exception handlers are responsible for similar tasks such logging and re-throwing.  </p><p>it will be easier to add an attribute such:   </p><p>[ExceptionHandlerAttribute(Type=&quot;ArgumentException&quot;, Handler=”HandlerFunctionName”)]  </p><p>[ExceptionHandlerAttribute(Type= &quot;FileNotFound&quot;, RetryCount=3, RetryDelay=”00:01”, Handler=”HandlerFunctionName”)]  </p><p>Let a binary rewriter add the try/catch with the proper code end exception type in each catch block.</p> <p>  
</p><p>  </p>

---
#### George Quiroga on 3/5/2020, 04:24 PM: 

<p>Something like this would be great. I spend a lot of time repeating similar try-catch and log exception handlers to much of my code. Having a shorter method of specifying handlers for exception processing would let me be more productive.</p>

#### Visual Studio Feedback System on 3/5/2020, 11:45 PM: 

Thank you for taking the time to provide your suggestion.  We will do some preliminary checks to make sure we can proceed further.  We'll provide an update once the issue has been triaged by the product team.

**Comment by gafter** [comment]
It was a decision very early in the C# language to depart from languages like Java that include exceptions in method contracts. It is something we're unlikely to revisit. However, I'm moving this issue to the language design repository in case there is further discussion.

**Comment by HaloFour** [comment]
This proposal seems less about checked exceptions and more about having a declarative way via attributes to declare exception handling, and having a "binary rewriter" that can implement that exception handling in your code.  Theoretically a use case for AOP-style source generators, although I feel that this approach is significantly kludgier than just writing `try`/`catch` blocks manually.
